File: test/int64_times_int64_test.c
#include "cint_arith.h"
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

// 測試向量結構: 保存 A, B, 以及預期結果 R
typedef struct {
    int64_t A;
    int64_t B;
    int128_t R;
} test_vector_t;

static test_vector_t test_vectors[] = {
    { 0xA06193792E3C31BEULL, 0x5172B98A41062CF3ULL, {{0xEC8D4C6FA423DF5AULL, 0xE19405CFE37A2C38ULL}} },
    { 0xA97A4AB8F1D70DB7ULL, 0x5A57A0C2CA121E87ULL, {{0x90CF8A7E18E1AD81ULL, 0xE177607F02F46173ULL}} },
    { 0x9BE5BB28894FD45FULL, 0xA8C8EEA7212EF4D3ULL, {{0x5C73B7FAAB48964DULL, 0x221A75D2825ADD23ULL}} },
    { 0x22C3C45E4663A118ULL, 0xFD656F55DE78EC15ULL, {{0xF8B74DFDF1EE56F8ULL, 0xFFA57B0DE9B9DE5AULL}} },
    { 0x68D0BCE3E89CDB01ULL, 0x37A03159D6177849ULL, {{0x35110B8A5E79EB49ULL, 0x16C66F3FC1CC79BFULL}} },
    { 0x7655D790639D2CD8ULL, 0x802B62C3B9812383ULL, {{0xD650ACA527677A88ULL, 0xC4E9224A591A588BULL}} },
    { 0xEB614FCD53A42AA8ULL, 0xF32659C021A04170ULL, {{0xA2D6E22B97A75180ULL, 0x0108F829C63FF31CULL}} },
    { 0x58E8B8022376D283ULL, 0x489E23B63774E1EAULL, {{0xDF243E5361FD8EBEULL, 0x19385FC94529025AULL}} },
    { 0xC982602D54285886ULL, 0xC9E5738EEB8CB8CEULL, {{0x4518549E235F8BD4ULL, 0x0B84265D70DA0D45ULL}} },
    { 0xC6790F139206833BULL, 0xC6FA78B6421A6F54ULL, {{0xD8A7E52D4707A45CULL, 0x0CD049AE144E4A27ULL}} },
    { 0x0000000000000000ULL, 0x0000000000000000ULL, {{0x0000000000000000ULL, 0x0000000000000000ULL}} },
    { 0x0000000000000001ULL, 0x0000000000000001ULL, {{0x0000000000000001ULL, 0x0000000000000000ULL}} },
    { 0xFFFFFFFFFFFFFFFFULL, 0xFFFFFFFFFFFFFFFFULL, {{0x0000000000000001ULL, 0x0000000000000000ULL}} },
    { 0x7FFFFFFFFFFFFFFFULL, 0x7FFFFFFFFFFFFFFFULL, {{0x0000000000000001ULL, 0x3FFFFFFFFFFFFFFFULL}} },
    { 0x8000000000000000ULL, 0x8000000000000000ULL, {{0x0000000000000000ULL, 0x4000000000000000ULL}} },
    { 0x7FFFFFFFFFFFFFFFULL, 0x8000000000000000ULL, {{0x8000000000000000ULL, 0xC000000000000000ULL}} },
};

int main(void)
{
    int errors = 0;
    int num_tests = sizeof(test_vectors)/sizeof(test_vectors[0]);
    int128_t result;

    for (int i = 0; i < num_tests; i++) {
        // 呼叫被測函式
        int64_times_int64(&test_vectors[i].A, &test_vectors[i].B, &result);

        // 驗證結果: 比對 v[0], v[1]
        if (result.v[0] != test_vectors[i].R.v[0] ||
            result.v[1] != test_vectors[i].R.v[1]) {
            errors++;
            printf("Test #%d FAILED!\n", i);
            printf("  A=%%lld, B=%%lld\n", (long long)test_vectors[i].A, (long long)test_vectors[i].B);
            printf("  Expected R: [%%016llX, %%016llX]\n",
                   (long long)test_vectors[i].R.v[1],
                   (long long)test_vectors[i].R.v[0]);
            printf("  Got R:      [%%016llX, %%016llX]\n",
                   (long long)result.v[1],
                   (long long)result.v[0]);
        }
    }

    if (errors == 0) {
        printf("All tests PASSED!\n");
    } else {
        printf("%%d tests FAILED.\n", errors);
    }

    return errors ? 1 : 0;
}

File: Archive/test/int64_times_int64_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import random

def int128_two_complement(x):
    """
    將 Python 的 int (可正可負，無限精度) 對應到 128-bit 的二補數表示。
    回傳一個非負整數 n，範圍在 [0, 2^128-1]。
    例如:
      若 x = 123，回傳 123
      若 x = -1，回傳 2^128 - 1
    """
    # 確保 x 落在 [-2^127, 2^127 - 1] 之間。
    if not (-2**127 <= x < 2**127):
        raise ValueError("超過 128-bit 帶正負號可表達範圍")

    # 若 x 為負，做二補數: x + 2^128
    if x < 0:
        x = (x + (1 << 128)) & ((1 << 128) - 1)
    return x

def int64_two_complement(x):
    """
    將 Python 的 int (可正可負，無限精度) 對應到 64-bit 的二補數表示。
    回傳一個非負整數 n，範圍在 [0, 2^64-1]。
    例如:
      若 x = 123，回傳 123
      若 x = -1，回傳 2^64 - 1
    """
    # 確保 x 落在 [-2^63, 2^63 - 1] 之間。
    if not (-2**63 <= x < 2**63):
        raise ValueError("超過 64-bit 帶正負號可表達範圍")

    # 若 x 為負，做二補數: x + 2^64
    if x < 0:
        x = (x + (1 << 64)) & ((1 << 64) - 1)
    return x



def split_uint128_into_two_64bit(n):
    """
    將 128-bit (非負) 的整數 n 分割成兩個 64-bit (v0, v1)。
    v0 為最低 64 bits，v1 為高位 64 bits。
    """
    v0 = n & ((1 << 64) - 1)
    v1 = n >> 64
    return v0, v1

# -----------------------------
# 主要程式開始
# -----------------------------

num_tests = 10  # 例: 產生 10 組隨機測試
test_vectors = []

# 1) 產生隨機測試案例
for _ in range(num_tests):
    A = random.randint(-2**63, 2**63 - 1)
    B = random.randint(-2**63, 2**63 - 1)
    R = A * B  # Python 自動做 64位*64位 => 任意精度

    # 確認結果在 [-2^127, 2^127-1] 之間
    if not (-2**127 <= R < 2**127):
        raise ValueError("R 超出 128-bit 帶正負號可表達範圍")

    test_vectors.append((A, B, R))

# 2) 也可手動加入一些特別的角落案例 (corner cases)
corner_cases = [
    (0, 0),
    (1, 1),
    (-1, -1),
    (2**63-1, 2**63-1),   # 9223372036854775807 * 9223372036854775807
    (-2**63, -2**63),    # -9223372036854775808 * -9223372036854775808
    (2**63-1, -2**63),   # ...
    # 其他您想測的特例
]

for (A, B) in corner_cases:
    R = A * B
    test_vectors.append((A, B, R))

# --------------------------------
# 產生 .c 檔
# --------------------------------

filename = "test/int64_times_int64_test.c"
with open(filename, "w") as f:
    # 先寫檔頭 (include、結構定義等)
    f.write('#include "cint_arith.h"\n')
    f.write('#include <stdio.h>\n')
    f.write('#include <stdint.h>\n')
    f.write('#include <stdlib.h>\n')
    f.write('\n')
    f.write('// 測試向量結構: 保存 A, B, 以及預期結果 R\n')
    f.write('typedef struct {\n')
    f.write('    int64_t A;\n')
    f.write('    int64_t B;\n')
    f.write('    int128_t R;\n')  # R 為 {v[0], v[1]}  (uint64_t)
    f.write('} test_vector_t;\n\n')

    # 開始定義測試陣列
    f.write('static test_vector_t test_vectors[] = {\n')

    for (A, B, R) in test_vectors:
        # 轉成二補數 128-bit
        R_128 = int128_two_complement(R)
        v0, v1 = split_uint128_into_two_64bit(R_128)
        A_64 = int64_two_complement(A)
        B_64 = int64_two_complement(B)

        # 將 R 寫成 {{0xXXXXXXXX, 0xXXXXXXXX}}
        # 注意 C99 語法: 用兩層大括號初始化 struct 中的陣列
        f.write('    { 0x%016XULL, 0x%016XULL, {{0x%016XULL, 0x%016XULL}} },\n' % (A_64, B_64, v0, v1))

    f.write('};\n\n')

    # 產生 main() 函式，逐筆測試
    f.write('int main(void)\n{\n')
    f.write('    int errors = 0;\n')
    f.write('    int num_tests = sizeof(test_vectors)/sizeof(test_vectors[0]);\n')
    f.write('    int128_t result;\n\n')
    f.write('    for (int i = 0; i < num_tests; i++) {\n')
    f.write('        // 呼叫被測函式\n')
    f.write('        int64_times_int64(&test_vectors[i].A, &test_vectors[i].B, &result);\n\n')
    f.write('        // 驗證結果: 比對 v[0], v[1]\n')
    f.write('        if (result.v[0] != test_vectors[i].R.v[0] ||\n')
    f.write('            result.v[1] != test_vectors[i].R.v[1]) {\n')
    f.write('            errors++;\n')
    f.write('            printf("Test #%d FAILED!\\n", i);\n')
    f.write('            printf("  A=%lld, B=%lld\\n", (long long)test_vectors[i].A, (long long)test_vectors[i].B);\n')
    f.write('            printf("  Expected R: [%016llX, %016llX]\\n",\n')
    f.write('                   (long long)test_vectors[i].R.v[1],\n')
    f.write('                   (long long)test_vectors[i].R.v[0]);\n')
    f.write('            printf("  Got R:      [%016llX, %016llX]\\n",\n')
    f.write('                   (long long)result.v[1],\n')
    f.write('                   (long long)result.v[0]);\n')
    f.write('        }\n')
    f.write('    }\n\n')
    f.write('    if (errors == 0) {\n')
    f.write('        printf("All tests PASSED!\\n");\n')
    f.write('    } else {\n')
    f.write('        printf("%d tests FAILED.\\n", errors);\n')
    f.write('    }\n\n')
    f.write('    return errors ? 1 : 0;\n')
    f.write('}\n')

print(f"已產生 C 測試檔案: {filename}")
